backend/app/services/api_sirene.rb

```diff
@@ -137,4 +125,28 @@ def codes_naf
   def categories_juridiques
     JSON.parse(File.read("./public/categories_juridiques_20200701.json"))
   end
+
+  def get_token
+    token_response = Http.instance.post({
+      url: "#{insee_host}/token",
+      body: {grant_type: "client_credentials"},
+      api_key: Base64.strict_encode64("#{insee_consumer_key}:#{insee_consumer_secret}"),
+      use_basic_auth_method: true,
+      use_form_content_type: true,
+      tag: "API Insee"
+    })
+    token_response.parse
+  end
+
+  def cached_access_token
+    cached_access_token = Rails.cache.read("insee/access_token")
+
+    if cached_access_token.nil?
+      token = get_token
+      Rails.cache.write("insee/access_token", token["access_token"], expires_in: token["expires_in"])
```

Comment:
This might fail when the access_token is about to expire.

please substract the expires_in with a `safe_renewal_interval` of 1 minute to prevent usage of an outdated cached token when the token is close to expiration.

Reply:
I've added the modification, Thank you ! :)